File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.rs

use std::{collections::*, hash::Hash};

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32>
    {
        let mut counts = nums.into_iter().count_unique();
        let mut result = Vec::new();

        for _ in 0..k {
            let Some((&num,_)) = counts.iter().max_by_key(|&(_,count)| count)
            else {
                unreachable!()
            };
            result.push(num);
            counts.remove(&num);
        }

        return result;
    }
}

pub trait CountUnique<T> {
    fn count_unique(self) -> HashMap<T,usize>;
}

impl<T,I> CountUnique<T> for I
where
    T: Eq + Hash,
    I: Iterator<Item=T>
{
    fn count_unique(self) -> HashMap<T,usize> {
        self.fold(HashMap::new(), |mut counts, elem| {
            *counts.entry(elem).or_default() += 1;
            counts
        })
    }
}

File: Data Structures & Algorithms/count-paths/submission-0.cpp

class Solution {
public:
    int uniquePaths(int m, int n)
    {
        if (m <= 0 || n <= 0) {
            return 0;
        }

        auto dp = vector<vector<int>>(m,vector<int>(n));
        dp[0][0] = 1;

        for (auto i = 1; i < m; ++i) {
            dp[i][0] = dp[i-1][0];
        }
        for (auto j = 1; j < n; ++j) {
            dp[0][j] = dp[0][j-1];
        }

        for (auto i = 1; i < m; ++i)
            for (auto j = 1; j < n; ++j)
        {
            dp[i][j] = dp[i-1][j] + dp[i][j-1];
        }

        return dp[m-1][n-1];
    }
};
